py/dml/ctree.py

```diff
@@ -3733,6 +3733,8 @@ def read(self):
                 + '(_traitref_t) {_object_vtables[__id.id], __id};})')
 
 def mkTraitUpcast(site, sub, parent):
+    if isinstance(sub, NonValue):
+        raise sub.exc()
```

Comment:
I could also place this check in `mkCast`, but saw no reason to prefer it there as opposed to here.

Reply:
This got me thinking a little about what sort of errors .exc() actually propagates up and whether they make sense here. As far as I can tell, it should be fine (though I am not entirely sure which nonvalues _can_ even appear here)